components/support/ktx/src/main/java/mozilla/components/support/ktx/kotlinx/coroutines/flow/Flow.kt

```diff
@@ -0,0 +1,34 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.support.ktx.kotlinx.coroutines.flow
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.filter
+
+/**
+ * Returns a [Flow] containing only values of the original [Flow] that have changed compared to
+ * the value emitted before them.
+ *
+ * Example:
+ * Original Flow: A, B, B, C, A, A, A, D, A
+ * Returned Flow: A, B, C, A, D, A
+ */
+fun <T> Flow<T>.ifChanged(): Flow<T> {
+    val lock = Any()
+    var observedValueOnce = false
+    var lastValue: T? = null
+
+    return filter { value ->
+        synchronized(lock) {
```

Comment:
I think I need a comment to understand why we need `synchronized` here :). In the docs it says "By default, flows are _sequential_ and all flow operations are executed sequentially in the same coroutine, with an exception for a few operations specifically designed to introduce concurrency into flow execution such as [buffer] and [flatMapMerge]. See their documentation for details.".

Reply:
My thinking here was that `observedValueOnce` and `lastValue` are getting created on the calling thread and then read inside the lambda running on some coroutine and that  this could cause a visibility issue. But I think you are right, since they are only read and written inside the coroutine lambda and if this happens in the same coroutine then I think we could strip that. 🤔

Reply:
Tests still pass without it. Updated the PR and removed the lock.